<div style="border:solid blue 2px; padding: 20px">
<font color='blue'>Привет, Александр!</font>

Твоё решение проверил <font color='blue'>Александр Цымбалов</font>

Если у тебя будут какие-то <font color='blue'>вопросы</font>, то обязательно их <font color='blue'>задавай</font>

Я буду обращаться к тебе на "ты", т.к. мы здесь все студенты)
<br><br>
Мои комментарии будут в отдельных ячейках <font color='blue'>Markdown</font> с заголовком «Комментарий ревьюера».

**Пожалуйста, не перемещай, не изменяй и не удаляй их**. 
Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Замечания, которые необходимо обработать, иначе я не смогу принять проект </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Небольшие замечания или вопросы по пониманию темы. Если есть желание и время, то у тебя есть возможность сделать проект ещё лучше </div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Хвалебные отзывы или другого рода комментарии)</div>

**Если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом.**

Мне будет <font color='blue'> легче </font> отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Лучше в таком формате, но как тебе <font color='blue'> удобно </font></div></font>

Версии комментариев я буду нумеровать

<br>

<div style="border:solid blue 2px; padding: 20px">
<font color='blue'>Итоги 1 ревью:</font>

Хорошая работа с текстом! <br>Получились правильные функции для очистки и лемматизации, корректное получение признаков TF-IDF и обучение моделей. В работе есть ошибки, которые стоит исправить:
    <ul>
        <li>Не используется предобработанный текст</li>
        <li>Модели не прошли минимальный порог по метрике</li>
    </ul>
Подробнее с ошибками ты сможешь познакомиться ниже в работе.<br>
Что я советую посмотреть в первую очередь, чтобы улучшить показатели моделей:
    <ul>
        <li>Внимательный подбор параметров для лог. регрессии (`C`) и леса (`n_estimators`)</li>
        <li>Посмотреть распределение в целевом признаке</li>
        <li>Посмотреть параметры `TfidfVectorizer`</li>
    </ul>
Успехов! Жду на ревью!

<div style="border:solid blue 2px; padding: 20px">
<font color='blue'>Итоги 2 ревью:</font>

Привет! <br>Теперь в работе верно используется лемматизированный текст для обучения, но не используется очистка.<br>По поводу метрики ниже минимального порога - предлагаю ответить на следующие вопросы: 
    <ul>
        <li>При подборе гиперпараметров учитываются ли базовые гиперпараметры? Насколько большой разброс относительно базового? <br>В случае лог. регрессии советую смотреть только на `C`</li>
        <li>Было ли сравнение `ngram_range=(1, 1)` и `ngram_range=(2, 2)` по финальной метрике при прочих равных?</li>
        <li>Стоит посмотреть на функцию `clear_text` и `TfidfVectorizer`</li>
    </ul>
С помощью лог. регрессии уже можно пройти порог по метрике, ты на финишной прямой! <br>Буду ждать на ревью. Удачи!

<div style="border:solid green 2px; padding: 20px">
<font color='green'>Итоги 3 ревью:</font>

Работа завершена успешно! <br>Правки позволили пройти минимальный порог по метрике. <br>Я добавил советы по поводу функции очистки, утечки тестовых данных в кросс-валидации, а также ответил на твои вопросы в конце работы, поэтому прошу ознакомиться. <br>Оставляю для тебя подборку полезных материалов, которые помогут глубже погрузиться в мир `Машинного обучения для текстов`:

<br>Полезные материалы:
    <ul><li>Про BERT</li>
    <ul>
        <li>Яндекс Практикум, RuBERT</li>
        <li>https://towardsdatascience.com/fine-tuning-bert-for-text-classification-54e7df642894 - Полный туториал-обучение по BERT</li>
        <li>https://medium.com/analytics-vidhya/text-classification-with-bert-using-transformers-for-long-text-inputs-f54833994dfd - Тоже туториал про классификации, но для длинных текстов</li>
        <li>https://huggingface.co/docs/transformers/tasks/sequence_classification - Официальный пример из документации huggingface</li>
        <li>https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/text_classification.ipynb - <b>Отличный туториал по реальному соревнованию</b></li>
    </ul>
    <li>Про GPU</li>
    <ul>
        <li>https://colab.research.google.com/ - Google Colab для эффективного обучения</li>
        <li>https://www.tutorialspoint.com/google_colab/google_colab_using_free_gpu.htm - Как включить GPU в Google Colab</li>
        <li>https://huggingface.co/docs/transformers/performance - Как BERT обучать на GPU</li>
    </ul>
            <li>Про catboost</li>
    <ul>
        <li>https://towardsdatascience.com/10x-times-fast-catboost-training-speed-with-an-nvidia-gpu-5ffefd9b57a6 - Сравнение GPU и CPU + код обучения на GPU</li>
        <li>https://colab.research.google.com/github/catboost/tutorials/blob/master/tools/google_colaboratory_cpu_vs_gpu_tutorial.ipynb - Тетрадка туториал по GPU и CPU catboost</li>
    </ul>
</ul>
    
Сейчас активно используются RNN (LSTM) и трансформеры (BERT, ELMO, GPT/2/3/n и др.). Они не являются панацеей, так как и TF-IDF или Word2Vec + модели из классического ML тоже могут решать задачи в текстах. \
BERT тяжелый, есть готовые модели, есть надстройки над библиотекой transformers. Если, обучать BERT на GPU (можно в Google Colab или Kaggle), то должно быть побыстрее.\
https://huggingface.co/transformers/model_doc/bert.html \
https://t.me/renat_alimbekov \
https://colah.github.io/posts/2015-08-Understanding-LSTMs/ - Про LSTM \
https://web.stanford.edu/~jurafsky/slp3/10.pdf - про энкодер-декодер модели, этеншены\
https://pytorch.org/tutorials/beginner/transformer_tutorial.html - официальный гайд
по трансформеру от создателей pytorch\
https://transformer.huggingface.co/ - поболтать с трансформером \
Библиотеки: allennlp, fairseq, transformers, tensorflow-text — множествореализованных
методов для трансформеров методов NLP \
Word2Vec https://radimrehurek.com/gensim/models/word2vec.html 

<font color='green'>Пример BERT с GPU:
```python
%%time
from tqdm import notebook
batch_size = 2 # для примера возьмем такой батч, где будет всего две строки датасета
embeddings = [] 
for i in notebook.tqdm(range(input_ids.shape[0] // batch_size)):
        batch = torch.LongTensor(input_ids[batch_size*i:batch_size*(i+1)]).cuda() # закидываем тензор на GPU
        attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).cuda()
        
        with torch.no_grad():
            model.cuda()
            batch_embeddings = model(batch, attention_mask=attention_mask_batch)
        
        embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy()) # перевод обратно на проц, чтобы в нумпай кинуть
        del batch
        del attention_mask_batch
        del batch_embeddings
        
features = np.concatenate(embeddings) 
```
Можно сделать предварительную проверку на наличие GPU.\
Например, так: ```device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")```\
Тогда вместо .cuda() нужно писать .to(device)</font>
    
Спасибо за работу и удачи на курсе, Александр!

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

In [1]:
import pandas as pd
import numpy as np
import nltk
import re
import matplotlib.pyplot as plt
import time

from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer 


from nltk.corpus import stopwords as nltk_stopwords
from nltk.stem import WordNetLemmatizer
 

nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('punkt')
stopwords = set(nltk_stopwords.words('english'))




[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /home/jovyan/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /home/jovyan/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [2]:
data = pd.read_csv('/datasets/toxic_comments.csv')
data.head(15)

text  toxic
0   Explanation\nWhy the edits made under my usern...      0
1   D'aww! He matches this background colour I'm s...      0
2   Hey man, I'm really not trying to edit war. It...      0
3   "\nMore\nI can't make any real suggestions on ...      0
4   You, sir, are my hero. Any chance you remember...      0
5   "\n\nCongratulations from me as well, use the ...      0
6        COCKSUCKER BEFORE YOU PISS AROUND ON MY WORK      1
7   Your vandalism to the Matt Shirvington article...      0
8   Sorry if the word 'nonsense' was offensive to ...      0
9   alignment on this subject and which are contra...      0
10  "\nFair use rationale for Image:Wonju.jpg\n\nT...      0
11  bbq \n\nbe a man and lets discuss it-maybe ove...      0
12  Hey... what is it..\n@ | talk .\nWhat is it......      1
13  Before you start throwing accusations and warn...      0
14  Oh, and the girl above started her arguments w...      0

In [3]:
data.info


<bound method DataFrame.info of                                                      text  toxic
0       Explanation\nWhy the edits made under my usern...      0
1       D'aww! He matches this background colour I'm s...      0
2       Hey man, I'm really not trying to edit war. It...      0
3       "\nMore\nI can't make any real suggestions on ...      0
4       You, sir, are my hero. Any chance you remember...      0
...                                                   ...    ...
159566  ":::::And for the second time of asking, when ...      0
159567  You should be ashamed of yourself \n\nThat is ...      0
159568  Spitzer \n\nUmm, theres no actual article for ...      0
159569  And it looks like it was actually you who put ...      0
159570  "\nAnd ... I really don't think you understand...      0

[159571 rows x 2 columns]>

<div class="alert alert-info">
<font color='black'> <b>  В датафрейме имеются 2 столбца: сам текст и целевой признак </font> <b>

In [4]:
display (data['toxic'].value_counts(sort = True))

0    143346
1     16225
Name: toxic, dtype: int64

In [5]:
data.duplicated().sum()

0

<div class="alert alert-info">
<font color='black'> <b>  Проверим данные на дубликаты: дубликаты отсутствуют </font> <b>

In [6]:
data.isna().sum()

text     0
toxic    0
dtype: int64

<div class="alert alert-info">
<font color='black'> <b>  Проверим данные на пропуски: пропуски отсутствуют </font> <b>

<div class="alert alert-success">
<b>1 Комментарий ревьюера ✔️:</b> Правильный вводный анализ датасета</div>


In [7]:
corpus = data['text']
 
def clear_text(text): 
    text = re.sub(r'[^a-zA-Z ]', ' ', str(text)) 
 
    return " ".join(text.split()) 
 
clear_corpus = corpus.apply(clear_text) 


<div class="alert alert-warning">
<b>2 Комментарий ревьюера ⚠️:</b> 
    
`clear_corpus` - не весь корпус, а только первая запись. Попробуй применить `.apply(clear_text)` к `data['text']`:
    
```python
clear_corpus = corpus.apply(clear_text)


<div class="alert alert-info">
<font color='black'> <b>  Применим очистку текста </font> <b>

<div class="alert alert-success">
<b>3 Комментарий ревьюера ✔️:</b> Верно!</div>


In [8]:
lemmatizer = WordNetLemmatizer() 
def lemmatize(text):
    lemmatize_text = []
    for sentence in corpus:
        word_list = nltk.word_tokenize(sentence)
        lemmatize_text.append(' '.join([lemmatizer.lemmatize(w) for w in word_list]))   
 
    return lemmatize_text 
 
clear_lem_corpus=lemmatize(clear_corpus)

<div class="alert alert-info">
<font color='black'> <b>  Применим лемматизацию. Данные подготовлены </font> <b>

<div class="alert alert-success">
<b>1 Комментарий ревьюера ✔️:</b> Функции для лемматизации и очистки правильные!</div>


<div class="alert alert-danger">
<b>1 Комментарий ревьюера ❌:</b> 

`corpus` содержит нелемматизированный неочищенный текст. Результаты работы функций не сохраняются
<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал </div>
</div>



<div class="alert alert-warning">
<b>2 Комментарий ревьюера ⚠️:</b> 
    
Не используешь очищенный текст. Работаешь только с лемматизированным. Посмотри внимательнее используется ли `text` в функции  </div>



<div class="alert alert-warning">
<b>3 Комментарий ревьюера ⚠️:</b> 
    
В функцию `lemmatize` передаешь `clear_text`с именем `text`, но внутри функции в цикле итерируешься по `corpus`, который не `clear_corpus`, т.е. снова не используешь очищенный текст, а только лемматизированный
</div>


## Обучение

<div class="alert alert-warning">
<b>3 Комментарий ревьюера ⚠️:</b> 
    
Немного про утечку тестовых данных (совет): если делать кросс-валидацию на признаках TF-IDF - произойдет утечка тестовых данных. Чтобы не попадала информация о словах из тест. выборки в тренировочную, мы обучаем `TfidfVectorizer` на тренировочной выборке, после чего трансформируем тест. выборку. <br>**НО!** Наша ситуация ниже: в `GridSearchCV` тренировочная выборка делится на 2-тренировочных фолда и 1-тестовый фолд, но при этом `TfidfVectorizer` обучался на 3-фолдах в сумме. Решение проблемы: [Pipelines](https://stackoverflow.com/questions/46010617/do-i-use-the-same-tfidf-vocabulary-in-k-fold-cross-validation) </div>


In [9]:
count_tf_idf = TfidfVectorizer(stop_words=stopwords, ngram_range=(1, 1))

<div class="alert alert-success">
<b>1 Комментарий ревьюера ✔️:</b> Здорово, что ты используешь стоп-слова!</div>


In [10]:
target=data['toxic']

features_train, features_sum, target_train, target_sum= train_test_split(clear_lem_corpus, target, test_size=0.4, random_state=12345)
features_val, features_test, target_val, target_test = train_test_split(features_sum, target_sum, test_size=0.5, random_state=12345)



<div class="alert alert-warning">
<b>1 Комментарий ревьюера ⚠️:</b> 
    
На валидацию и тестирование лучше отдать по 20% от датасета (как рекомендуется в курсе)
<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал </div></div>


<div class="alert alert-success">
<b>2 Комментарий ревьюера ✔️:</b> Отлично!</div>


In [11]:
tf_idf_train = count_tf_idf.fit_transform(features_train)
tf_idf_val = count_tf_idf.transform(features_val)
tf_idf_test = count_tf_idf.transform(features_test)

<div class="alert alert-success">
<b>1 Комментарий ревьюера ✔️:</b> Не допустил утечку тестовых данных!</div>


In [12]:
%%time
model_rf = RandomForestClassifier(random_state=12345)
parameters_rf = {'n_estimators':list(range(1, 30)),"max_depth": list(range(4, 8))}
rf_grid = GridSearchCV(estimator=model_rf, param_grid = parameters_rf, cv = 3, n_jobs=-1, verbose=10)
rf_grid.fit(tf_idf_train, target_train)
params_rf = rf_grid.best_params_
rf_best = RandomForestClassifier(**params_rf, random_state=12345)
rf_best.fit(tf_idf_train, target_train)
predictions_rf = rf_best.predict(tf_idf_val)

Fitting 3 folds for each of 116 candidates, totalling 348 fits
[CV 1/3; 1/116] START max_depth=4, n_estimators=1...............................
[CV 1/3; 1/116] END .............max_depth=4, n_estimators=1; total time=   0.2s
[CV 2/3; 1/116] START max_depth=4, n_estimators=1...............................
[CV 2/3; 1/116] END .............max_depth=4, n_estimators=1; total time=   0.2s
[CV 3/3; 1/116] START max_depth=4, n_estimators=1...............................
[CV 3/3; 1/116] END .............max_depth=4, n_estimators=1; total time=   0.2s
[CV 1/3; 2/116] START max_depth=4, n_estimators=2...............................
[CV 1/3; 2/116] END .............max_depth=4, n_estimators=2; total time=   0.2s
[CV 2/3; 2/116] START max_depth=4, n_estimators=2...............................
[CV 2/3; 2/116] END .............max_depth=4, n_estimators=2; total time=   0.2s
[CV 3/3; 2/116] START max_depth=4, n_estimators=2...............................
[CV 3/3; 2/116] END .............max_depth=4, 


<div class="alert alert-warning">
<b>1 Комментарий ревьюера ⚠️:</b> 
    
Советую в первую очередь подбирать `n_estimators`
<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал </div>
</div></div>


<div class="alert alert-success">
<b>2 Комментарий ревьюера ✔️:</b> Верно!</div>


In [13]:
display (params_rf)

{'max_depth': 7, 'n_estimators': 1}

In [14]:
f1_score(target_val, predictions_rf)

0.005447941888619855

<div class="alert alert-info">
<font color='black'> <b>  При использовании модели случайного  леса метрика f1 получена далёкой от требуемой </font> <b>

In [15]:
%%time
model_lf = LogisticRegression(random_state=12345, class_weight='balanced')
parameters_lf = {"C": list(range(2,5))}
lf_grid = GridSearchCV(estimator=model_lf, param_grid = parameters_lf, cv = 3, n_jobs=-1, verbose=10)
lf_grid.fit(tf_idf_train, target_train)
params_lf = lf_grid.best_params_
lf_best = LogisticRegression(**params_lf, random_state=12345)
lf_best.fit(tf_idf_train, target_train)
predictions_lf = lf_best.predict(tf_idf_val)

Fitting 3 folds for each of 3 candidates, totalling 9 fits
[CV 1/3; 1/3] START C=2.........................................................


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[CV 1/3; 1/3] END .......................................C=2; total time=  42.4s
[CV 2/3; 1/3] START C=2.........................................................
[CV 2/3; 1/3] END .......................................C=2; total time=  24.8s
[CV 3/3; 1/3] START C=2.........................................................


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[CV 3/3; 1/3] END .......................................C=2; total time=  45.4s
[CV 1/3; 2/3] START C=3.........................................................


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[CV 1/3; 2/3] END .......................................C=3; total time=  42.6s
[CV 2/3; 2/3] START C=3.........................................................
[CV 2/3; 2/3] END .......................................C=3; total time=  29.1s
[CV 3/3; 2/3] START C=3.........................................................
[CV 3/3; 2/3] END .......................................C=3; total time=  30.7s
[CV 1/3; 3/3] START C=4.........................................................


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[CV 1/3; 3/3] END .......................................C=4; total time=  43.6s
[CV 2/3; 3/3] START C=4.........................................................


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[CV 2/3; 3/3] END .......................................C=4; total time=  43.1s
[CV 3/3; 3/3] START C=4.........................................................


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


[CV 3/3; 3/3] END .......................................C=4; total time=  42.6s


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


CPU times: user 3min 6s, sys: 4min 10s, total: 7min 16s
Wall time: 7min 17s


/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [16]:
display (params_lf)

{'C': 4}

In [17]:
f1_score(target_val, predictions_lf)

0.7641875220303137

<div class="alert alert-info">
<font color='black'> <b>  При использовании модели логистической регрессии метрика f1 получена в пределах допустимого значения. Применим эту модель на тестовых данных </font> <b>

<div class="alert alert-success">
<b>3 Комментарий ревьюера ✔️:</b> Отличный результат!</div>


In [19]:
predictions_lf_test = lf_best.predict(tf_idf_test)

In [20]:
f1_score(target_test, predictions_lf_test)

0.7563896336014299

<div class="alert alert-info">
<font color='black'> <b>  При использовании модели логистической регрессии на тестовых данных метрика f1 получена в пределах допустимого значения - 0.756 </font> <b>

<div class="alert alert-success">
<b>3 Комментарий ревьюера ✔️:</b> Порог пройден!</div>


## Выводы

<div class="alert alert-info">
<font color='black'> <b>  Лучше всего в решении задачи показала себя модель логистической регрессии. Метрика f1 в пределах допустимого </font> <b>


<div class="alert alert-warning">
<b>1 Комментарий ревьюера ⚠️:</b> По поводу нуля спорно. Лог. регрессия показывает неплохие результаты. Остальное в итогах 1 ревью 
<div class="alert alert-info"> <b>Комментарий студента:</b> Вроде учел все советы, но ничего не вышло(( где главная ошибка? (0.3 далеко от 0.75) Нужно применить "upsample"/"downsample"? Еще что-то? Приходится ждать работу моделей около часа в сумме</div>
</div>



<div class="alert alert-warning">
<b>2 Комментарий ревьюера ⚠️:</b> В итогах 2 ревью
 <div class="alert alert-info"> <b>Комментарий студента:</b> Достиг требуемого результата! Пара вопросов: 1) Лемматизацию я применяю не ко всему тексту. Это корректно? (если применять apply, то считает очень долго и ядро через час умирает) 2) Расскажи пожалуйста про параметр "С" логистической регрессии. Не совсем понял, что от него зависит</div>

<div class="alert alert-warning">
<b>3 Комментарий ревьюера ⚠️:</b>

<ol>
<li> Правильно применять твой способ лемматизацию не к тексту целиком, а ко словам, поэтому всё верно</li>
<li> <b>C</b> - это обратная сила регуляризации. Меньшие значения определяют более сильную регуляризацию.
    Очень советую посмотреть про регуляризацию <a href="https://neerc.ifmo.ru/wiki/index.php?title=%D0%A0%D0%B5%D0%B3%D1%83%D0%BB%D1%8F%D1%80%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D1%8F">здесь</a>, т.к. тут есть пример с лог. регрессией</li>
</ol>

## Чек-лист проверки

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Данные загружены и подготовлены
- [x]  Модели обучены
- [x]  Значение метрики *F1* не меньше 0.75
- [x]  Выводы написаны